File: janMarkusMilan/assignments/slidingwindow/lc340/lc340.js
// Problem Statement #

// Given a string, find the length of the longest substring in it with no more than K distinct characters.

// You can assume that K is less than or equal to the length of the given string.

/*
Hashmap: Keep track of every instance of each character
Var: max
Loop (forms a window):
  add the char to the hashmap (if it is in hashmap, increment)
  if occurrence of a char > K:
    slide window
  update length to max

return the max

'araaci'
{}
Iteration 1: {'a': 1}
Iteration 2: {'a': 1, 'r': 1}
Iteration 3: {'a': 2, 'r': 1}
Iteration 4: {'a': 3, 'r': 1}
Iteration 5: {'a': 3, 'r': 1, 'c': 1}
Iteration 6: {'a': 3, 'r': 1, 'c': 1, 'i': 1}
*/

function longest_substring_with_k_distinct(str, k) {
  // TODO: Write code here
  let max = 0;
  const charMap = {};

  // Form a 'window' using two pointers over a portion of the data
  let leftPtr = 0;
  for (let rightPtr = 0; rightPtr < str.length; rightPtr++) {
    // Get the current char and check if it is in the hashmap (will be undefined or null if it isn't)
    const currentChar = charMap[str[rightPtr]];

    // If it is present, increment the value by 1, else add it to the hashmap
    if (currentChar >= 0) {
      charMap[str[rightPtr]]++;
    } else {
      charMap[str[rightPtr]] = 1;
    }

    // If the occurrence (value) of the current char is greater than k, slide the window over
    while (charMap[str[rightPtr]] > k) {
      // Get the left-most char in the window
      const leftChar = str[leftPtr];

      // Decrement the value of the char in the hashmap by 1
      charMap[leftChar]--;

      // Increment the left pointer by 1 (update the beginning position of the window)
      leftPtr++;
    }

    // Keep track of the max of each iteration (window size will increase at the end of each iteration)
    max = Math.max(max, rightPtr - leftPtr + 1);
  }

  return max;
}

console.log(
  `Length of the longest substring: ${longest_substring_with_k_distinct(
    "araaci",
    2
  )}`
);
console.log(
  `Length of the longest substring: ${longest_substring_with_k_distinct(
    "araaci",
    1
  )}`
);
console.log(
  `Length of the longest substring: ${longest_substring_with_k_distinct(
    "cbbebi",
    3
  )}`
);

// Solution
// -----
// let windowStart = 0,
// maxLength = 0,
// charFrequency = {};

// // in the following loop we'll try to extend the range [window_start, window_end]
// for (let windowEnd = 0; windowEnd < str.length; windowEnd++) {
// const rightChar = str[windowEnd];
// if (!(rightChar in charFrequency)) {
//   charFrequency[rightChar] = 0;
// }
// charFrequency[rightChar] += 1;
// // shrink the sliding window, until we are left with 'k' distinct characters in the char_frequency
// while (Object.keys(charFrequency).length > k) {
//   const leftChar = str[windowStart];
//   charFrequency[leftChar] -= 1;
//   if (charFrequency[leftChar] === 0) {
//     delete charFrequency[leftChar];
//   }
//   windowStart += 1; // shrink the window
// }
// // remember the maximum length so far
// maxLength = Math.max(maxLength, windowEnd - windowStart + 1);
// }

// return maxLength;

File: xiaqianZhang/assignments/slidingwindow/lc340/lc340.py
# Problem Statement #

# Given a string, find the length of the longest substring in it with no more than K distinct characters.
# You can assume that K is less than or equal to the length of the given string.
'''
input: string and k that is target unique characters
output: return longest length of substring that contains target k

initialize windowStart = 0 start at index 0, resultLen = 0
initliaze a directionary to count the current appearence of the unique characters
for each index value in the string:
  check if the current unique characters is in the dictionary:
  if not append the unique characters into the dictionary by set its count as 0
  else if the character is in the dict then increment its count by 1

  After we are getting the numbers of unique charater's count in the directionary:
    while the unique characters in the uniqueV is greater than k distinct required characters:
      initlize the unique charater for the windowStart index, starts with index 0's value
      decrement the count of the value that position at windowStart index
      if the count of the windoeStart index is 0:
        then delete the key from the dictionary
      
      increment the windowStart by 1
    after this while loop, then update the longest length
return resultLen

'''
def longest_substring_with_k_distinct(str1, k):
  windowStart = 0
  resultLen = 0
  uniqueV = {}
  for each in range(len(str1)):
    currentV = str1[each]
    if currentV not in uniqueV:
      uniqueV[currentV] = 0
    uniqueV[currentV]+=1
    
    while len(uniqueV) > k:
      startV = str1[windowStart]
      uniqueV[startV] -= 1
      if uniqueV[startV] == 0:
        del uniqueV[startV]
      
      windowStart+=1
    
    resultLen = max(resultLen, each - windowStart + 1)
  return resultLen

def main():
  print("Length of the longest substring: " + str(longest_substring_with_k_distinct("araaci", 2)))
  # "araa" length of 4
  # araa -> raa -> aac -> ac -> ci
  print("Length of the longest substring: " + str(longest_substring_with_k_distinct("araaci", 1)))
  # "aa" length of 2
  # a -> r -> aa -> c -> i
  print("Length of the longest substring: " + str(longest_substring_with_k_distinct("cbbebi", 3)))
  # "cbbeb" "bbebi" length of 5
  # cbbeb -> bbebi -> bebi -> ebi -> bi -> i


main()


















# Solution
# -----
  # window_start = 0
  # max_length = 0
  # char_frequency = {}

  # # in the following loop we'll try to extend the range [window_start, window_end]
  # for window_end in range(len(str1)):
  #   right_char = str1[window_end]
  #   if right_char not in char_frequency:
  #     char_frequency[right_char] = 0
  #   char_frequency[right_char] += 1

  #   # shrink the sliding window, until we are left with 'k' distinct characters in the char_frequency
  #   while len(char_frequency) > k:
  #     left_char = str1[window_start]
  #     char_frequency[left_char] -= 1
  #     if char_frequency[left_char] == 0:
  #       del char_frequency[left_char]
  #     window_start += 1  # shrink the window
  #   # remember the maximum length so far
  #   max_length = max(max_length, window_end-window_start + 1)
  # return max_length
